src/test/ui/associated-types/defaults-specialization.rs

```diff
@@ -0,0 +1,47 @@
+// compile-fail
+
+#![feature(associated_type_defaults, specialization)]
+
+trait Tr {
+    type Ty = u8;
+
+    fn make() -> Self::Ty;
+}
+
+struct A<T>(T);
+// In a `default impl`, assoc. types are defaulted as well,
+// so their values can't be assumed.
+default impl<T> Tr for A<T> {
+    fn make() -> u8 { 0 }
```

Comment:
Would also be good to add 2 more tests `A2` and `B2` to check that if you write `-> Self::Ty { expr_assumes_default_type }` then that too is rejected.